core/src/providers/tiktoken/tiktoken.rs

```diff
@@ -478,6 +489,24 @@ impl CoreBPE {
     }
 
     fn _encode_native(&self, text: &str, allowed_special: &HashSet<&str>) -> (Vec<usize>, usize) {
+        // Try regex-based encoding first, fall back to byte-based on stack overflow
+        match std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+            self._encode_native_internal(text, allowed_special)
+        })) {
+            Ok(result) => result,
+            Err(_) => {
+                // Stack overflow or panic occurred, use byte-based encoding
+                eprintln!("Warning: Regex encoding failed, falling back to byte-based encoding");
```

Comment:
Why not just error in the API with a clear error code? 

Also aren't we timing out before we reach the stack overflow?